File: CameraWebServer_bluethoot/SmileDetector.cpp
#include "SmileDetector.h"

SmileDetector::SmileDetector() {
  counter = 0;
}

void SmileDetector::no_face() {
  this->save_smile_counter = 0;
  corners_distances.clear();
  lips_distances.clear();
  counter = 0;
}

void SmileDetector::face_detected(std::vector<std::pair<int, int>> face) {
  if (corners_distances.size() >= MAX_FRAMES) {
    corners_distances.pop_back();
    lips_distances.pop_back();
  }

  corners_distances.insert(corners_distances.begin(), distance(face[LEFT_CORNER], face[RIGHT_CORNER]) /
                           distance(face[TOP_NORM], face[BOTTOM_NORM]));
  lips_distances.insert(lips_distances.begin(), distance(face[TOP_LIP], face[BOTTOM_LIP]) /
                        distance(face[TOP_NORM], face[BOTTOM_NORM]));
  counter++;
}

bool SmileDetector::is_smile() {
  if (counter < MAX_FRAMES) {
    return false;
  }
  if ( this->save_smile_counter > 0)
  {
    this->save_smile_counter--;
    std::cout << "good smile" << std::endl;
    return true;
  }
  double corner_dist_after = mean(corners_distances.begin(), corners_distances.begin() + (MAX_FRAMES / 2));
  double corner_dist_before = mean(corners_distances.begin() + (MAX_FRAMES / 2) + 1, corners_distances.end());

  double lips_dist_after = mean(lips_distances.begin(), lips_distances.begin() + (MAX_FRAMES / 2));
  double lips_dist_before = mean(lips_distances.begin() + (MAX_FRAMES / 2) + 1, lips_distances.end());

  if (2 * lips_dist_before > lips_dist_after) {
    std::cout << "not enough teethes" << std::endl;
    return false;
  }

  if (1.2 * corner_dist_before > corner_dist_after) {
    std::cout << "not wide enough b:" << corner_dist_before << " a:" << corner_dist_after << std::endl;
    return false;
  }

  std::cout << "good smile" << std::endl;
  this->save_smile_counter = 10;
  return true;
}

double SmileDetector::distance(std::pair<int, int> point1, std::pair<int, int> point2) {
  return std::sqrt(std::pow(point1.first - point2.first, 2) + std::pow(point1.second - point2.second, 2));
}

double SmileDetector::mean(std::vector<int>::iterator begin, std::vector<int>::iterator end) {
  double sum = 0.0;
  int count = 0;

  for (auto it = begin; it != end; ++it) {
    sum += *it;
    count++;
  }

  if (count > 0) {
    return sum / count;
  } else {
    return 0.0;
  }
}

File: Project_Interface/SmileDetector.py
import math
import numpy as np


def _distance_(point1, point2):
    return math.sqrt((point1[0] - point2[0]) ** 2 + (point1[1] - point2[1]) ** 2)


class SmileDetector:
    def __init__(self):

        self.corners_distances = []
        self.lips_distances = []
        self.counter = 0
        self.MAX_FRAMES = 10
        self.LEFT_CORNER = 61
        self.RIGHT_CORNER = 291
        self.TOP_LIP = 13
        self.BOTTOM_LIP = 15
        self.LEFT_NORM = 44
        self.RIGHT_NORM = 274
        self.NOSE = 1
        self.CHIN = 152

    def no_face(self):
        self.corners_distances = []
        self.lips_distances = []
        self.counter = 0

    def face_detected(self, face):
        if len(self.corners_distances) >= self.MAX_FRAMES:
            self.corners_distances.pop()
            self.lips_distances.pop()

        self.corners_distances.insert(0, _distance_(face[self.LEFT_CORNER], face[self.RIGHT_CORNER]) /
                                      _distance_(face[self.LEFT_NORM], face[self.RIGHT_NORM]))
        self.lips_distances.insert(0, _distance_(face[self.TOP_LIP], face[self.BOTTOM_LIP]) /
                                   _distance_(face[self.LEFT_NORM], face[self.RIGHT_NORM]))
        self.counter += 1

    def is_smile(self):
        if self.counter < self.MAX_FRAMES:
            return False

        corner_dist_after = np.mean(self.corners_distances[0:int(self.MAX_FRAMES / 2)])
        corner_dist_before = np.mean(self.corners_distances[int(self.MAX_FRAMES / 2) + 1: self.MAX_FRAMES])

        lips_dist_after = np.mean(self.lips_distances[0:int(self.MAX_FRAMES / 2)])
        lips_dist_before = np.mean(self.lips_distances[int(self.MAX_FRAMES / 2) + 1: self.MAX_FRAMES])

        if 1.5 * lips_dist_before > lips_dist_after:
            print("not enough teethes")
            return False

        if 1.3 * corner_dist_before > corner_dist_after:
            print("not wide enough b:" + str(corner_dist_before) + " a:" + str(corner_dist_after))
            return False

        print("good smile")
        return True
